llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp

```diff
@@ -1737,6 +1737,42 @@ bool SYCLLowerESIMDPass::prepareForAlwaysInliner(Module &M) {
     F.addFnAttr(llvm::Attribute::NoInline);
   };
 
+  bool ModuleContainsGenXVolatile =
+      std::any_of(M.global_begin(), M.global_end(), [](const auto &Global) {
+        return Global.hasAttribute("genx_volatile");
+      });
+
+  auto requiresInlining = [=](Function &F) {
+    // If there are any genx_volatile globals in the module, inline
+    // noinline functions because load/store semantics are not valid for
+    // these globals and we cannot know for sure if the load/store target
+    // is one of these globals without inlining.
+    if (ModuleContainsGenXVolatile)
```

Comment:
Please check that this function does not overwrite NoInline for `assert` function. Otherwise, it cannot be lowered by GPU BE.
See it here: https://github.com/intel/llvm/blob/46af2115c7db20ef7c343e844b8669a78c16f574/llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp#L1801

Reply:
Thanks, I just added a lit test to lock this down

Reply:
I think the test is not testing the situation that is caused by this `if (ModuleContainsGenXVolatile) return true;` return of true because that module/test does not have volatile, right?

To me it looks like there is chance that L1802 would set the attribute to 'NoInline' for assert, and then the L1813 would set it to `AlwaysInline`.

Reply:
There's a continue on line 1803, so it wouldn't, but let me add a test.

Reply:
just pushed the new test

Reply:
Right, I missed that 'continue' statement. Thank you for the extra tests.